### ISSUE TITLE
Missing resource ID

### ISSUE THREAD

**Issue by yperess** [JakeWharton/butterknife#974]
I'm running:
* Android Studio 2.3.3
* Kotlin 1.1.2-3
* Butterknife 8.6.0

My project has an `app` module an a `library` module, in the app module I've included:

```
dependencies {
    ...
    compile "com.jakewharton:butterknife:$butterknife_version"
    kapt "com.jakewharton:butterknife-compiler:$butterknife_version"
}
```

In the library module I have:

```
buildscript {
    repositories {
        jcenter()
    }
    dependencies {
        classpath "com.jakewharton:butterknife-gradle-plugin:$butterknife_version"
    }
}

apply plugin: 'com.android.library'
apply plugin: 'com.jakewharton.butterknife'
apply plugin: 'kotlin-android'
apply plugin: 'kotlin-android-extensions'
apply plugin: 'kotlin-kapt'
android {
    ...
}
dependencies {
    ...
    compile "com.jakewharton:butterknife:$butterknife_version"
    kapt "com.jakewharton:butterknife-compiler:$butterknife_version"
}
```

My library code includes a custom view called CameraControlView with the following:

```
@OnClick(R2.id.switch_camera)
fun switchCamera() {
}
```

The generated `CameraControlView_ViewBinding.java` has the following offending line:

```
view = Utils.findRequiredView(source, 2131493059, "method 'switchCamera'");
```

R2.id.switch_camera is 0x7f0c00c3 (2131493059)
R.id.switch_camera is 0x7f0c00c3 in library
R.id.switch_camera is 0x7f100186 in app

Am I configuring butterknife wrong for the library or is this a bug?

**Comment by JakeWharton** [comment]
I doubt Butter Knife works in Kotlin libraries. That should emit an R
reference not a raw integer.

**Comment by yperess** [comment]
Do you mean that R2 should contain something like `switch_camera = com.example.R.id.switch_camera`?

The generated R2 contains:

```
    @IdRes
    public static final int switch_camera = 0x7f0c00c3;
```

Library R contains:

```
        public static int switch_camera=0x7f0c00c3;
```

App R contains:

```
        public static final int switch_camera=0x7f100186;
```

**Comment by JakeWharton** [comment]
No, the generated view binder should emit an R reference and not a raw
integer. R2 should have integers.

**Comment by yperess** [comment]
I see, any more info I can provide to debug?

**Comment by JakeWharton** [comment]
I'm pretty sure this flat out can't work in kapt because we use a
javac-internal API currently to look up the R class. I'd have to reach out
to the Kotlin people about it.

**Comment by yperess** [comment]
https://github.com/JetBrains/kotlin-examples/tree/master/gradle/android-butterknife

**Comment by yperess** [comment]
The above project when compiled generates a ViewBinding class that also uses constants instead of references.

**Comment by JakeWharton** [comment]
And it probably fails at runtime because of this.

**Comment by yperess** [comment]
Actually it runs, I imagine if we add more modules or names with conflicts it would fail and it just by luck managed to merge the resources without modifying the values.

**Comment by vanniktech** [comment]
I also just hit this. The problem is that the `R.id.myId` and `R2.id.myId` are different. 

I've also tried out the [example](https://github.com/JetBrains/kotlin-examples) and it does work surprisingly. @yperess have you found a way to make it work for your application?

**Comment by yperess** [comment]
@vanniktech No I haven't, I've been busy dealing with other things about our project so I just put a hold on using butterknife on our library modules for now. In theory what you could do is create a second pass. Attach a gradle task to run after generateReleaseSources (when the R.java file is created) then go through the .java files in the butterknife output directory and find all the constants that are also in the R.java file and replace them with references. It's sloppy but in theory it should work.

**Comment by RileyGB** [comment]
I have run in to this issue as well and have found a workaround. If you add a **Java** class and bind any resource in the same package which you are binding, then the resources in your Kotlin classes will resolve correctly.

For example, I include a java class called BindingHelper.java in every package where I bind resources in Kotlin with the following code:

`public class BindingHelper {
    @BindString(R2.string.app_name) String appName;
}`

**Comment by JakeWharton** [comment]
The values are meant to be different between R and R2, by the way.

**Comment by athornz** [comment]
We've been using butterknife in kotlin library modules and its been working great.

But after updating to 9.0.0-SNAPSHOT we've run into this issue. 

Are there changes in 9.0.0 that mean this will never work?

**Comment by yperess** [comment]
The guys at JetBrains just closed the ticket causing this. It's scheduled to be resolved in kapt 1.2.20

**Comment by athornz** [comment]
Oh great! Do you have a link to that ticket?

**Comment by yperess** [comment]
https://youtrack.jetbrains.com/issue/KT-18791

**Comment by JakeWharton** [comment]
So... no action to take?

**Comment by yperess** [comment]
From what I understand we wait. As soon as 1.2.20 is out I'll confirm and close this issue or update it.

**Comment by JakeWharton** [comment]
Thanks!

**Comment by RiteshChandnani** [comment]
I just updated to early access preview of Kotlin (version 1.2.20-eap-11) on Android studio and that binding issue is gone :)

**Comment by yperess** [comment]
That's correct, it does appear to be working in eap-11 and eap-33. @JakeWharton my vote is to leave this open until 1.2.20 goes into the main stream so that people can easily find adding the eap as a workaround until then, but totally your call.

**Comment by kylannjohnson** [comment]
For more fodder, it seems like this issue causes a lint error in AGP 3.1-alpha8.  If we remove the @BindView(R2.id.some_id) from Kotlin classes, then the lint error goes away.  We are also on Kotlin 1.2.20.

The warning given is just: `Expected resource of type id`

We'll either have to remove Butterknife or ignore for now.

**Comment by MyDogTom** [comment]
`OnClick` annotation is not working for me in library module. Is that the same problem as described here or should I create a separate issue?

ButterKnife: 9.0.0-SNAPSHOT
Kotlin: 1.2.40

For the following class
```kotlin
class MyFragment : BaseFragment() {

    @BindView(R2.id.textview_upsell_confirmation_title) lateinit var headlineTextView: TextView
    @BindView(R2.id.textview_upsell_confirmation_text) lateinit var confirmationTextView: TextView

    @OnClick(R2.id.button_premium_upsell_confirmation)
    internal fun upsellConfirmationClicked() {
    }
}
```

ButterKnife generates
```java
  @UiThread
  public MyFragment_ViewBinding(
      final MyFragment target, View source) {
    this.target = target;

    View view;
    target.headlineTextView = Utils.findRequiredViewAsType(source, R.id.textview_upsell_confirmation_title, "field 'headlineTextView'", TextView.class);
    target.confirmationTextView = Utils.findRequiredViewAsType(source, R.id.textview_upsell_confirmation_text, "field 'confirmationTextView'", TextView.class);
    view = Utils.findRequiredView(source, 2131493117, "method 'upsellConfirmationClicked$core_app_debug'");
    view7f0c00fd = view;
    view.setOnClickListener(new DebouncingOnClickListener() {
      @Override
      public void doClick(View p0) {
      }
    });
  }

```
At runtime I'm receiving `Caused by: android.content.res.Resources$NotFoundException: Unable to find resource ID`

I noticed, that for `@BindView` generated code contains reference to `R.id`. But for `@OnClick` generated code contains hardcoded constant.

**Comment by idanatz** [comment]
Having the same issue as @MyDogTom 
OnClick annotation generated code contains a reference to R.id. But for @OnClick generated code contains hardcoded constant and sometimes even a wrong R.id.

ButterKnife: 8.8.1
Kotlin: 1.2.50

**Comment by yperess** [comment]
Have you tried butterknife 9.0.0-rc1?

**Comment by frett** [comment]
It appears this issue was re-introduced with Kotlin 1.3.10.

R2 usage in a Kotlin class works correct with the following dependency versions:
ButterKnife: 9.0.0-rc1
Kotlin: 1.2.71

Upgrading Kotlin to 1.3.10 causes the R2 resource to be replaced with a raw int and not the correct R id in the generated ViewBinding class

**Comment by JakeWharton** [comment]
It's fixed in 1.3.20

**Comment by oliverspryn** [comment]
Good to know, waiting for 1.3.20 to hit stable before retrying butterknife. Using Kotlin extensions for the time being.

**Comment by oliverspryn** [comment]
[1.3.20 is out](https://blog.jetbrains.com/kotlin/2019/01/kotlin-1-3-20-released/)  🎉  Didn't get a chance to check out Butterknife again, though

**Comment by marcosalis** [comment]
I have just tested Kotlin 1.3.20 (stable) with Butterknife 9.0.0 on library modules, and everything seems to be working fine! 👍

**Comment by JakeWharton** [comment]
Glad to hear it!

**Comment by ysyyork** [comment]
> I have just tested Kotlin 1.3.20 (stable) with Butterknife 9.0.0 on library modules, and everything seems to be working fine! 👍

I just test the same kotlin 1.3.20 and butterknife 9.0.0 but the issue is still there? Are you sure about this?

**Comment by kkmike999** [comment]
the problem still here. 

FragmentA
```
@BindView(R2.id.itemView)
lateinit var itemView: ViewGroup

@OnClick(R2.id.itemView)
fun onItemClick(view: View) {}
```

FragmentA_ViewBinding
```
View view;
view = Utils.findRequiredView(source, R.id.itemView, "field 'itemView' and method 'onItemClick'");
```

if i kon't write `@BindView(R2.id.itemView)`,then `FragmentA_ViewBinding` will be:

```
View view;
view = Utils.findRequiredView(source, 2131427425, "method 'onItemClick'");
```
**where is the `R.id.itemView`??**

throw exception:
```
java.lang.IllegalStateException: Required view 'md_dialog_progress' with ID 2131427425 for method 'onItemClick' was not found. If this view is optional add '@Nullable' (fields) or '@Optional' (methods) annotation.
        at butterknife.internal.Utils.findRequiredView(Utils.java:88)
        at com.badambiz.live.fragment.FragmentA_ViewBinding.<init>(FragmentA_ViewBinding.java:24)
```

`2131427425` is not equals `R.id.itemView`

----
```
classpath 'com.android.tools.build:gradle:3.4.0'
classpath "org.jetbrains.kotlin:kotlin-gradle-plugin:1.3.50"
classpath "com.jakewharton:butterknife-gradle-plugin:9.0.0"
```

```
apply plugin: 'com.jakewharton.butterknife'
apply plugin: 'kotlin-android'
apply plugin: 'kotlin-android-extensions'
apply plugin: 'kotlin-kapt'

dependencies{
    api 'com.jakewharton:butterknife:9.0.0'
    kapt 'com.jakewharton:butterknife-compiler:9.0.0'
}
```

**Comment by newtalentxp** [comment]
the problem is still here, not fixed, fuck

**Comment by ysyyork** [comment]
@newtalentxp wowow easy man

**Comment by vy12021** [comment]
Hi, guys above.

[Stubs in kapt ](https://blog.jetbrains.com/kotlin/2015/06/better-annotation-processing-supporting-stubs-in-kapt/)

After my analysis, I found that some annotation meta information was erased during the generation of the java stub file. In our question, the annotations on the member variables retained the reference information, but the annotation information on the member methods was replaced with constant. This will cause the generated ButterKnife file to find the symbol index of the id if no member variable with the same id is defined. I guess this may be a bug in kapt that generates stub files. What is doubtful is that there is a difference in the processing of the two member elements.

The default path of kapt stub : 
`build/tmp/kapt3/stubs/debug/com.xxx.package/xxx.Java`

![20200221161442](https://user-images.githubusercontent.com/7430792/75016416-e8d84700-54c5-11ea-98a0-72aa08c78710.jpg)

@JakeWharton
